src/Microsoft.OData.Core/ODataBatchReader.cs

```diff
@@ -204,48 +247,167 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
+        protected abstract ODataBatchReaderState ReadAtChangesetStartImplementation();
+
+        /// <summary>
+        /// Implementation of the reader logic when in state 'ChangesetEnd'.
+        /// </summary>
+        /// <returns>The batch reader state after the read.</returns>
+        protected abstract ODataBatchReaderState ReadAtChangesetEndImplementation();
+
+        /// <summary>
+        /// Instantiate an <see cref="ODataBatchOperationRequestMessage"/> instance.
+        /// </summary>
+        /// <param name="streamCreatorFunc">The function for stream creation.</param>
+        /// <param name="method">The HTTP method used for this request message.</param>
+        /// <param name="requestUri">The request Url for this request message.</param>
+        /// <param name="headers">The headers for this request message.</param>
+        /// <param name="contentId">The contentId of this request message.</param>
+        /// <param name="dependsOnRequestIds">
+        /// The prerequisite request Ids of this request message. For batch in Json format,
+        /// some of these request Ids are resolved from prerequisite atomic groups that are
+        /// specified in the dependsOn attribute of the request.</param>
+        /// <returns>The <see cref="ODataBatchOperationRequestMessage"/> instance.</returns>
+        protected ODataBatchOperationRequestMessage BuildOperationRequestMessage(
+            Func<Stream> streamCreatorFunc,
+            string method,
+            Uri requestUri,
+            ODataBatchOperationHeaders headers,
+            string contentId,
+            IEnumerable<string> dependsOnRequestIds)
+        {
+            if (dependsOnRequestIds != null)
             {
-                case ODataBatchReaderState.Initial:
-                    // If we find an end boundary when in state 'Initial' it means that we
-                    // have an empty batch. The next state will be 'Completed'.
-                    return ODataBatchReaderState.Completed;
+                foreach (string id in dependsOnRequestIds)
+                {
+                    if (!this.payloadUriConverter.ContainsContentId(id))
+                    {
+                        throw new ODataException(Strings.ODataBatchReader_DependsOnIdNotFound(id, contentId));
+                    }
+                }
 
-                case ODataBatchReaderState.Operation:
-                    // If we find an end boundary in state 'Operation' we have finished
-                    // processing an operation and found the end boundary of either the
-                    // current changeset or the batch.
-                    return this.batchStream.ChangeSetBoundary == null
-                        ? ODataBatchReaderState.Completed
-                        : ODataBatchReaderState.ChangesetEnd;
+                ODataBatchUtils.ValidateReferenceUri(requestUri, dependsOnRequestIds,
+                    this.inputContext);

```

Comment:
>this.inputContext [](start = 20, length = 17)

See comment on ValidateReferenceUri in ODataBatchUtils.  This should just take the this.inputContext.BaseUri, and the logic in ODataBatchUtils should not depend on the format.

Reply:
Updated to use  this.inputContext.BaseUri directly. Also use the batchFormat value to indicate whether this is from MultipartMixed batch, in which case the reference uri validate should be skipped if the dependsOnIds is not specified (value is null) for backward compatibility with original Multipart/Mixed request. 

---
In reply to: [155353476](https://github.com/OData/odata.net/pull/992#discussion_r155353476) [](ancestors = 155353476)

Reply:
Still don't like passing in format, but will cover that in a separate comment.

---
In reply to: [156197925](https://github.com/OData/odata.net/pull/992#discussion_r156197925) [](ancestors = 156197925,155353476)